File: Spektra AI/edidio_spektra_ai/discovery.py
"""UDP discovery of eDIDIO controllers on the local network.

Python port of the JS engine's discovery probe (used by the REST gateway and the
Discord bot). Protocol matches the SpektraPlus app:

  - Send the probe byte "D" to UDP port 30303.
  - Target the DIRECTED SUBNET BROADCAST address of every active IPv4 interface
    (ip | ~netmask), since many networks don't route the global 255.255.255.255.
  - Controllers reply with JSON {NAME,MAC,IP,TLS,FW_VER,LINES} or a legacy
    "NAME\r\nMAC" text payload. We ignore our own "D" echo.
  - `TLS: true` means the unit accepts a TLS connection on port 443 (older units
    are plain TCP on port 23).
"""

from __future__ import annotations

import ipaddress
import json
import socket
import time

DISCOVERY_PORT = 30303
PROBE = b"D"

LINE_LABELS = {0: "Empty", 1: "DALI", 2: "DMX", 3: "DMX-In", 4: "Auto"}


def _broadcast_targets() -> list[str]:
    """Directed broadcast address for every active IPv4 interface."""
    targets: set[str] = set()
    hostname = socket.gethostname()
    try:
        addrs = socket.getaddrinfo(hostname, None, socket.AF_INET)
    except socket.gaierror:
        addrs = []

    for _family, _type, _proto, _canon, sockaddr in addrs:
        ip = sockaddr[0]
        if ip.startswith("127."):
            continue
        # Assume a /24 when we can't read the real netmask from stdlib; this is
        # the common case on lighting LANs. Users can also pass --host directly.
        try:
            net = ipaddress.ip_network(f"{ip}/24", strict=False)
            targets.add(str(net.broadcast_address))
        except ValueError:
            continue

    targets.add("255.255.255.255")  # fallback / same-subnet
    return sorted(targets)


def _parse(text: str, ip: str) -> dict | None:
    text = text.strip()
    if text == "D":
        return None  # our own echo
    if text.startswith("{") and text.endswith("}"):
        try:
            data = json.loads(text)
            data["IP"] = ip
            return data
        except json.JSONDecodeError:
            return None
    parts = text.split("\r\n")
    if len(parts) >= 2:
        return {"NAME": parts[0], "MAC": parts[1], "IP": ip}
    return None


def summarize_lines(lines) -> str:
    """Compact 'Line 1: DALI · Lines 2-4: DMX' summary."""
    if not isinstance(lines, list) or not lines:
        return "Unknown"
    parts = []
    start = 0
    for i in range(1, len(lines) + 1):
        if i == len(lines) or lines[i] != lines[start]:
            label = LINE_LABELS.get(lines[start], f"Type {lines[start]}")
            if start == i - 1:
                parts.append(f"Line {start + 1}: {label}")
            else:
                parts.append(f"Lines {start + 1}-{i}: {label}")
            start = i
    return " · ".join(parts)


def discover(timeout: float = 2.0) -> list[dict]:
    """Broadcast a probe and collect controller replies until `timeout`."""
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
    sock.settimeout(0.5)

    found: dict[str, dict] = {}
    try:
        sock.bind(("0.0.0.0", 0))
        for target in _broadcast_targets():
            try:
                sock.sendto(PROBE, (target, DISCOVERY_PORT))
            except OSError:
                pass  # a single unreachable target shouldn't abort discovery

        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            try:
                data, addr = sock.recvfrom(1024)
            except socket.timeout:
                continue
            except OSError:
                break
            info = _parse(data.decode("utf-8", errors="replace"), addr[0])
            if info and addr[0] not in found:
                found[addr[0]] = info
    finally:
        sock.close()

    return list(found.values())


if __name__ == "__main__":
    devices = discover()
    if not devices:
        print("No eDIDIO controllers found on the local network.")
    for d in devices:
        print(f"{d.get('NAME', '?')}  {d.get('IP')}  MAC={d.get('MAC')}  "
              f"TLS={d.get('TLS')}  FW={d.get('FW_VER')}  "
              f"[{summarize_lines(d.get('LINES'))}]")

File: Note-RED Package/lib/discovery.js
// Discover eDIDIO controllers on the local network via UDP broadcast.
//
// Protocol (matches the SpektraPlus desktop app):
//   - Send the probe string "D" to UDP port 30303.
//   - Send it to the DIRECTED SUBNET BROADCAST address of every active network
//     interface (IP | ~netmask), NOT the global 255.255.255.255 — many networks
//     don't route the global broadcast, so the directed one is what reaches the
//     controller.
//   - Controllers reply with either JSON {NAME,MAC,IP,TLS,FW_VER,LINES} or a
//     legacy "NAME\r\nMAC" text payload. We ignore our own "D" echo.
//   - Newer firmware includes `TLS: true`, meaning the unit accepts a TLS
//     connection on port 443 (older units are plain TCP on port 23).

const dgram = require('node:dgram');
const os = require('node:os');

const DISCOVERY_PORT = 30303;
const PROBE = Buffer.from('D', 'utf8');

// VPN heuristic (mirrors SpektraPlus): on a 172.x interface, also unicast to a
// small range of likely controller addresses since broadcast may not traverse.
const VPN_PREFIX = '172.';
const VPN_UNICAST_START = 8;
const VPN_UNICAST_END = 13;

// Build the list of target IPs to probe: the directed broadcast for every
// active IPv4 interface, plus VPN unicast targets where applicable.
function getDiscoveryTargets() {
	const targets = new Set();
	const interfaces = os.networkInterfaces();

	for (const name of Object.keys(interfaces)) {
		for (const iface of interfaces[name] || []) {
			if (iface.family !== 'IPv4' || iface.internal) continue;

			const ip = iface.address.split('.').map(Number);
			const mask = iface.netmask.split('.').map(Number);
			// Broadcast address = IP OR (NOT netmask)
			const broadcast = ip.map((part, i) => part | (~mask[i] & 255)).join('.');
			targets.add(broadcast);

			if (iface.address.startsWith(VPN_PREFIX)) {
				const base = `${ip[0]}.${ip[1]}.${ip[2]}`;
				for (let i = VPN_UNICAST_START; i <= VPN_UNICAST_END; i++) {
					const target = `${base}.${i}`;
					if (target !== iface.address) targets.add(target);
				}
			}
		}
	}

	// Fallback if no interfaces were enumerable.
	if (targets.size === 0) targets.add('255.255.255.255');
	return [...targets];
}

// Parse a single UDP response payload into a controller info object, or null.
function parseResponse(text, ip) {
	if (text === 'D') return null; // our own echo

	if (text.startsWith('{') && text.endsWith('}')) {
		try {
			const json = JSON.parse(text);
			return { ...json, IP: ip };
		} catch {
			return null;
		}
	}

	// Legacy format: "NAME\r\nMAC"
	const parts = text.split('\r\n');
	if (parts.length >= 2) {
		return { NAME: parts[0], MAC: parts[1], IP: ip };
	}
	return null;
}

// Returns a Promise resolving to an array of controller info objects:
//   { NAME, MAC, IP, TLS?, FW_VER?, LINES? }
function discoverControllers(timeoutMs = 2000) {
	return new Promise((resolve) => {
		const socket = dgram.createSocket({ type: 'udp4', reuseAddr: true });
		const found = new Map(); // ip -> info
		let done = false;

		const finish = () => {
			if (done) return;
			done = true;
			try {
				socket.close();
			} catch {
				/* already closed */
			}
			resolve([...found.values()]);
		};

		socket.on('error', finish);

		socket.on('message', (msg, rinfo) => {
			const info = parseResponse(msg.toString('utf8').trim(), rinfo.address);
			if (info && !found.has(rinfo.address)) found.set(rinfo.address, info);
		});

		socket.on('listening', () => {
			socket.setBroadcast(true);
			for (const target of getDiscoveryTargets()) {
				socket.send(PROBE, DISCOVERY_PORT, target, (err) => {
					// A single unreachable target shouldn't abort discovery.
					if (err) console.warn(`[discovery] send to ${target} failed: ${err.message}`);
				});
			}
			setTimeout(finish, timeoutMs);
		});

		// Bind to the discovery port on all interfaces so controllers that reply
		// to the source port reach us.
		socket.bind(DISCOVERY_PORT, '0.0.0.0');
	});
}

module.exports = { discoverControllers, DISCOVERY_PORT };
